tests/contracts/ValidationRules.sol

```diff
@@ -22,65 +25,97 @@ contract Dummy {
+    function eq(string memory a, string memory b) internal pure returns (bool) {
+        return keccak256(bytes(a)) == keccak256(bytes(b));
     }
 
-
-    function funTLOAD() external returns(uint256) {
-        uint256 tval;
-        assembly {
-            tval := tload(0)
+    function startsWith(string memory str, string memory prefix) internal pure returns (bool) {
+        bytes memory strBytes = bytes(str);
+        bytes memory prefixBytes = bytes(prefix);
+        if (prefixBytes.length > strBytes.length) {
+            return false;
         }
-        emit State(tval, tval);
-        return tval;
-    }
-
-    event State(uint oldState, uint newState);
-
-    function setState(uint _state) public {
-        emit State(state, _state);
-        state = _state;
-    }
-
-    function revertOOG() public {
-        uint256 i = 0;
-        while(true) {
-            keccak256(abi.encode(i++));
+        for (uint256 i = 0; i < prefixBytes.length; i++) {
+            if (strBytes[i] != prefixBytes[i]) {
+                return false;
+            }
         }
+        return true;
     }
 
-    function revertOOGSSTORE() public {
-        state = state;
+    function slice(string memory str, uint256 start, uint256 length) internal pure returns (string memory) {
+        bytes memory strBytes = bytes(str);
+        require(start + length <= strBytes.length, "Out of bounds");
+        bytes memory result = new bytes(length);
+        for (uint256 i = 0; i < length; i++) {
+            result[i] = strBytes[start + i];
+        }
+        return string(result);
     }
-}
-
-library ValidationRules {
 
-    function eq(string memory a, string memory b) internal pure returns (bool) {
-        return keccak256(bytes(a)) == keccak256(bytes(b));
+    function includes(string memory str, string memory substr) internal pure returns (bool) {
+        bytes memory strBytes = bytes(str);
+        bytes memory substrBytes = bytes(substr);
+        if (substrBytes.length > strBytes.length) {
+            return false;
+        }
+        for (uint256 i = 0; i <= strBytes.length - substrBytes.length; i++) {
+            bool matchFound = true;
+            for (uint256 j = 0; j < substrBytes.length; j++) {
+                if (strBytes[i + j] != substrBytes[j]) {
+                    matchFound = false;
+                    break;
+                }
+            }
+            if (matchFound) {
+                return true;
+            }
+        }
+        return false;
     }
 
     //return by runRule if string is unknown.
     uint constant public UNKNOWN = type(uint).max;
 
     error CustomError(string error, uint256 code);
 
-    function runRule(string memory rule, IState account, TestCoin coin, ValidationRulesStorage self) internal returns (uint) {
+    function runFactorySpecificRule(uint nonce, string memory rule, address _entryPoint, address create2address) internal {
+        if (eq(rule, "EXTCODEx_CALLx_undeployed_sender")) {
```

Comment:
didn't find a test where "EXTCODEx_CALLx_undeployed_sender" is used

Reply:
Had a similar problems with `EXTCODEx_CALLx_undeployed_sender` - it is giving a command to thefactory to make an action **before** the sender is deployed. There is only one place like that in our code and it had to be handled by a separate function, so I created the `runFactorySpecificRule` for it.

It is used here:
https://github.com/eth-infinitism/bundler-spec-tests/blob/a0b2bd7474672e278b24c4e4a478feb298c45c3e/tests/single/bundle/test_storage_rules.py#L437